### PR TITLE
Fix snapshot density matrix

### LINKED ISSUE #303: DensityMatrix snapshot extension
### What is the expected behavior?

This will be used by the density matrix method of the QasmSimulator.

```python
class SnapshotDensityMatrix(Snapshot):
    __init__(self, label, num_qubits):
        super(label, "density_matrix", num_qubits, 0, None)
 

# Monkey patch function
def snapshot_density_matrix(self, label, qubits=None):
     # see base snapshot monkey patch for converting qubits to
     # be all qubits in circuit

circuit.snapshot_density_matrix = snapshot_density_matrix
```

The `num_qubits` parameter is actually ignored by Aer (it is always an all-qubit instruction), however it is used by Terra to make the snapshot act as a barrier.

### PR DESCRIPTION
**Pull request by chriseclectic**
### Summary

Fixes density matrix snapshot and adds tests.

Closes #303 
~Blocked by #355~

### Details and comments

* Changes signature density matrix snapshot circuit function to not need qubit number, if it is not specified it will act on all qubits.
* Fixes qasm simulator to default to density matrix method if there is a density matrix snapshot (previously it would raise an exception of unsupported instruction for statevector simulator).
* Fixes density matrix to be an average snapshot instead of single shot snapshot
* Addes tests for density matrix extension and simulator implementation.

### CODE REVIEW

File: src/simulators/qasm/qasm_controller.hpp [hunk 1]
```diff
@@ -527,10 +539,16 @@ QasmController::simulation_method(const Circuit &circ,
...
     // If we didn't select extended stabilizer above proceed to the default switch clause
     default: {
-      // Default method is statevector
+      // For default we use statevector followed by density matrix (for the case
+      // when the circuit contains invalid instructions for statevector)
+      if (validate_state(Statevector::State<>(), circ, noise_model, false)) {
+        return Method::statevector;
+      }
+      // If circuit contains invalid instructions for statevector throw a hail
+      // mary and try for density matrix.
```
Comment (atilag) [inline src/simulators/qasm/qasm_controller.hpp]: Ave María!!

File: test/terra/backends/qasm_simulator/qasm_snapshot.py [hunk 2]
```diff
@@ -30,6 +30,10 @@
     snapshot_state_pre_measure_statevector_nondeterministic,
     snapshot_state_post_measure_statevector_nondeterministic)
 
+def get_snapshots(data, label, snapshot_type):
+        """Format snapshots as list of Numpy arrays"""
```
Comment (atilag) [inline test/terra/backends/qasm_simulator/qasm_snapshot.py]: I think this has one more level of indentation, right?

File: src/simulators/qasm/qasm_controller.hpp [hunk 3]
```diff
@@ -468,6 +468,18 @@ QasmController::simulation_method(const Circuit &circ,
                                   bool validate) const {
   // Check simulation method and validate state
   switch(simulation_method_) {
+    case Method::statevector: {
+      if (validate) {
+        if (simulation_precision_ == Precision::single_precision) {
+          Statevector::State<QV::QubitVector<float>> state; 
+          validate_state(state, circ, noise_model, true);
```
[the remark this answers is not in the document]
Comment (chriseclectic) [reply src/simulators/qasm/qasm_controller.hpp]: the state arg is actually different (template param <float> or <double> in each branch);

File: test/terra/backends/qasm_simulator/qasm_snapshot.py (hunk 2 above)
Comment (chriseclectic) [reply test/terra/backends/qasm_simulator/qasm_snapshot.py]: indeed!

Comment (chriseclectic) [reply test/terra/backends/qasm_simulator/qasm_snapshot.py]: Will merge here and fix in #380

**Pull request merged**